src/internal/document_api/mod.rs

```diff
@@ -632,6 +637,36 @@ async fn resolve_keys_for_grants(
     ))
 }
 
+/// Get a cached policy or run the given Future to get the evaluated policy from the webservice.
+/// Policies that evaluate cleanly with no invalid users or groups are cached for future use.
+async fn get_cached_policy_or<F>(
+    config: &PolicyCachingConfig,
+    grant: &PolicyGrant,
+    policy_cache: &PolicyCache,
+    get_policy_f: F,
+) -> Result<(Vec<DocAccessEditErr>, Vec<WithKey<UserOrGroup>>), IronOxideErr>
+where
+    F: Future<Output = Result<PolicyResponse, IronOxideErr>>,
+{
+    // if there's a value in the cache, use it
+    if let Some(cached_policy) = policy_cache.get(grant) {
+        Ok((vec![], cached_policy.clone()))
+    } else {
+        // otherwise query the webservice and cache the result if there are no errors
+        get_policy_f.await.map(|policy_resp| {
+            let (errs, public_keys) = process_policy(&policy_resp);
+            if errs.is_empty() {
+                //if the cache has grown too large, clear it prior to adding new entries
+                if policy_cache.len() >= config.max_entries {
+                    policy_cache.clear()
```

Comment:
I'm a bit sad that this is the invalidation strat we're using. Even removing some arbitrary entry would be better, wouldn't it? Also, you didn't check to see if the if the insert would be actually replacing an entry (concurrent requests).

Reply:
I agree that it's pretty simple minded. All we're really trying to guarantee here is that the cache can't grow without bound for a long running process. I feel like anything smarter probably should include time or usage stats. 

I did consider doing what you suggest and just throwing out a single, random entry, which seems reasonable, but then we need to introduce some machinery to decide _which_ (random) entry to throw out and it just didn't seem worth it.